dcos_installer/test_backend.py

```diff
@@ -67,6 +67,24 @@ def test_set_superuser_password(tmpdir):
         assert passlib.hash.sha512_crypt.verify('foo', config['superuser_password_hash'])
 
 
+def test_generate_node_upgrade_script(tmpdir):
+    """
+    Test the generate upgrade script works
+    """
+    config_extra = "\nmaster_list: ['10.0.0.1', '10.0.0.2', '10.0.0.5']\n"
+
+    with tmpdir.as_cwd():
+        genconf_dir = tmpdir.join('genconf')
+        genconf_dir.ensure(dir=True)
+        make_default_config_if_needed('genconf/config.yaml', config_extra)
+        genconf_dir.join('ip-detect').write('#!/bin/bash\necho 127.0.0.1')
+        tmpdir.join('artifacts/bootstrap/12345.bootstrap.tar.xz').write('contents_of_bootstrap', ensure=True)
+        tmpdir.join('artifacts/bootstrap/12345.active.json').write('{"active": "contents"}', ensure=True)
+
+    output = subprocess.check_output(['dcos_installer', '--generate-node-upgrade-script', '1.8.7'], cwd=str(tmpdir))
+    assert output.decode('ascii').strip('\n').split("Node upgrade script URL: ", 1)[1].endswith("dcos_node_upgrade.sh")
```

Comment:
It'd be safer to decode this as UTF-8. It might be even saferer to not decode the output at all and compare against the literal bytes.

Let's also enforce that the URL is on the very last line of the output, to make sure users don't have to sift through text to get to it. I think this does verify that, but indirectly and in a way that'll fail with an unclear `IndexError` if the URL line prefix isn't there. Looks like we can split by line and then compare against the last line.